hstream/src/HStream/Server/Handler.hs

```diff
@@ -402,84 +403,110 @@ subscribeHandler :: ServerContext
                  -> ServerRequest 'Normal SubscribeRequest SubscribeResponse
                  -> IO (ServerResponse 'Normal SubscribeResponse)
 subscribeHandler ServerContext{..} (ServerNormalRequest _metadata SubscribeRequest{..}) = do
-  es' <- mapM getSourceConnector subscribeRequestSubscriptions
-  case V.all isRight es' of
-    False -> do
-      let resp = SubscribeResponse subscribeRequestSubscriptionId (Enumerated $ Right HStreamServerErrorUnknownError)
-      return (ServerNormalResponse resp [] StatusUnknown "")
-    True  -> do
-      let connectorsWithStreamName = fromRight undefined <$> es'
-      atomicModifyIORef' subscribedConnectors (\hm -> (HM.insert subscribeRequestSubscriptionId connectorsWithStreamName hm, ()))
-      let resp = SubscribeResponse subscribeRequestSubscriptionId (Enumerated $ Right HStreamServerErrorNoError)
-      return (ServerNormalResponse resp [] StatusOk "")
+  es' <- loop subscribeRequestSubscriptions scLDClient 0 (length subscribeRequestSubscriptions) V.empty
+  let resp = SubscribeResponse subscribeRequestSubscriptionId
+  case es' of
+    Left err  -> do
+      return $ ServerNormalResponse resp [] StatusInternal $ StatusDetails (C.pack . displayException $ err)
+    Right res -> do
+      atomicModifyIORef' subscribedReaders (\hm -> (HM.insert subscribeRequestSubscriptionId res hm, ()))
+      return $ ServerNormalResponse resp [] StatusOk ""
   where
-    getSourceConnector :: StreamSubscription -> IO (Either SomeException (TL.Text,SourceConnector))
-    getSourceConnector StreamSubscription{..} = do
+    loop :: V.Vector StreamSubscription
+         -> LDClient
+         -> Int -> Int  -- current index and the length of StreamSubscription
+         -> V.Vector (TL.Text, LDSyncCkpReader)
+         -> IO (Either SomeException (V.Vector (TL.Text, LDSyncCkpReader)))
+    loop opts client idx len vec
```

Comment:
one subscription should map to one reader and the reader should subscribe all streams in this subscription.

Reply:
fixed